File: pangu-application/src/sslcert/sslcert.rs
use async_trait::async_trait;
use derive_more::Display;
use pangu_domain::{
    errors::Error,
    model::{DnsProvider, SSLCertificate, DnsProviderType},
};
use serde::{Deserialize, Serialize};
#[derive(Debug, Clone, Serialize, Deserialize, Display)]
#[display(
    fmt = "SSLCertRequest: [domain={},mail={}, dns_provider={}",
    domain,
    mail,
    provider_id
)]
pub struct SSLCertRequest {
    pub provider_id: i64,
    pub mail: String,
    pub domain: String,
    pub subdomain: String,
    pub sn: String,
    // pub dns_provider: String,
}
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DnsProviderCreateReq {
    pub name: String,
    pub api_key: String,
    pub api_secret: String,
    pub provider_type: DnsProviderType,
}
#[async_trait]
pub trait SSLCertApplicationService {
    async fn add_dns_provider(&self, provider: DnsProviderCreateReq) -> Result<i64, Error>;
    async fn create_cert(&self, cert: SSLCertRequest) -> Result<(), Error>;
    async fn list_dns_providers(&self) -> Result<Vec<DnsProvider>, Error>;
    async fn list_sslcerts(&self) -> Result<Vec<SSLCertificate>, Error>;
    async fn get_sslcert_by_sn(&self, sn: &str) -> Result<SSLCertificate, Error>;
    async fn remove_sslcert(&self, id: i64) -> Result<(), Error>;
}

File: pangu-studio/src/api/sslcert.ts
import { P } from '@tauri-apps/api/event-30ea0228'
import { invoke } from '@tauri-apps/api/tauri'

export interface DnsProvider {
  id: number
  name: string
  api_key: string
  api_secret: string
  create_time: string
  update_time: string
}

export interface SSLCertificate {
  id: number
  domains: string
  cert_chain: string
  private_key: string
  status: string
  deleted: boolean
  addition: string
  create_time: string
  update_time: string
}
export interface SSLCertificateCreate {
  sn: string
  provider_id: number
  mail: string
  domain: string
  subdomain: string
}

export interface CreateDnsProvider {
  name: string
  api_key: string
  api_secret: string
  provider_type: string
}

export async function listDnsProvider(): Promise<DnsProvider[]> {
  return invoke('list_dns_providers')
}
export async function listSSLCertificate(): Promise<SSLCertificate[]> {
  return invoke('list_sslcerts')
}
export async function applyCertificate(
  cert: SSLCertificateCreate
): Promise<SSLCertificate> {
  return invoke('apply_certificate', { cert })
}
export async function getSslCertBySN(sn: string): Promise<SSLCertificate> {
  return invoke('get_sslcert_by_sn', { sn })
}
export async function removeSslCert(id: number): Promise<SSLCertificate> {
  return invoke('remove_sslcert', { id })
}

export async function addDnsProvider(
  provider: CreateDnsProvider
): Promise<DnsProvider> {
  return invoke('add_dns_provider', { provider })
}
